nav2_costmap_2d/plugins/costmap_filters/binary_filter.cpp

```diff
@@ -0,0 +1,310 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/binary_filter.hpp"
+
+#include <cmath>
+#include <utility>
+#include <memory>
+#include <string>
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "geometry_msgs/msg/point_stamped.hpp"
+
+#include "nav2_costmap_2d/costmap_filters/filter_values.hpp"
+#include "nav2_util/occ_grid_values.hpp"
+
+namespace nav2_costmap_2d
+{
+
+BinaryFilter::BinaryFilter()
+: filter_info_sub_(nullptr), mask_sub_(nullptr),
+  binary_state_pub_(nullptr), filter_mask_(nullptr), mask_frame_(""), global_frame_(""),
+  binary_state_(false)
+{
+}
+
+void BinaryFilter::initializeFilter(
+  const std::string & filter_info_topic)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Declare "binary_state_topic" parameter specific to BinaryFilter only
+  std::string binary_state_topic;
+  declareParameter("binary_state_topic", rclcpp::ParameterValue("binary_state"));
+  node->get_parameter(name_ + "." + "binary_state_topic", binary_state_topic);
+
+  filter_info_topic_ = filter_info_topic;
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    logger_,
+    "BinaryFilter: Subscribing to \"%s\" topic for filter info...",
+    filter_info_topic_.c_str());
+  filter_info_sub_ = node->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    filter_info_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&BinaryFilter::filterInfoCallback, this, std::placeholders::_1));
+
+  // Get global frame required for binary state publisher
+  global_frame_ = layered_costmap_->getGlobalFrameID();
+
+  // Create new binary state publisher
+  binary_state_pub_ = node->create_publisher<nav2_msgs::msg::BinaryState>(
+    binary_state_topic, rclcpp::QoS(10));
+  binary_state_pub_->on_activate();
+
+  // Initialize state as "false" by-default
+  changeState(false);
```

Comment:
I think there should be a param for the default value so that we can initialize / reset by user requested behavior.

Reply:
Added default (and parametrized by user) binary filter state in https://github.com/ros-planning/navigation2/pull/3228/commits/056ed5e09f0092855591bed2223511fc8d0cac71